pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+  of protocols. As discussed in `rationale`_, we propose to add static support
+  for such behavior. In addition, to allow users to achieve such runtime
+  behavior for *user defined* protocols a special ``@runtime`` decorator will
+  be provided, see detailed `discussion`_ below.
+
+* TypeScript [typescript]_ provides support for user defined classes and
+  interfaces. Explicit implementation declaration is not required and
+  structural subtyping is verified statically. For example::
+
+    interface LabeledItem {
+        label: string;
+        size?: int;
+    }
+
+    function printLabel(obj: LabeledValue) {
+        console.log(obj.label);
+    }
+
+    let myObj = {size: 10, label: "Size 10 Object"};
+    printLabel(myObj);
+
+  Note that optional interface members are supported. Also, TypeScript
+  prohibits redundant members in implementations. While the idea of
+  optional members looks interesting, it would complicate this proposal and
+  it is not clear how useful it will be. Therefore it is proposed to postpone
+  this; see `rejected`_ ideas. In general, the idea of static protocol
+  checking without runtime implications looks reasonable, and basically
+  this proposal follows the same line.
+
+* Go [golang]_ uses a more radical approach and makes interfaces the primary
+  way to provide type information. Also, assignments are used to explicitly
+  ensure implementation::
+
+    type SomeInterface interface {
+        SomeMethod() ([]byte, error)
+    }
+
+    if _, ok := someval.(SomeInterface); ok {
+        fmt.Printf("value implements some interface")
+    }
+
+  Both these ideas are questionable in the context of this proposal. See
+  the section on `rejected`_ ideas.
+
+
+.. _specification:
+
+Specification
+=============
+
+Terminology
+-----------
+
+We propose to use the term *protocols* for types supporting structural
+subtyping. The reason is that the term *iterator protocol*,
+for example, is widely understood in the community, and coming up with
+a new term for this concept in a statically typed context would just create
+confusion.
+
+This has the drawback that the term *protocol* becomes overloaded with
+two subtly different meanings: the first is the traditional, well-known but
+slightly fuzzy concept of protocols such as iterator; the second is the more
+explicitly defined concept of protocols in statically typed code.
+The distinction is not important most of the time, and in other
+cases we propose to just add a qualifier such as *protocol classes*
+when referring to the static type concept.
+
+If a class includes a protocol in its MRO, the class is called
+an *explicit* subclass of the protocol. If a class is a structural subtype
+of a protocol, it is said to implement the protocol and to be compatible
+with a protocol. If a class is compatible with a protocol but the protocol
+is not included in the MRO, the class is an *implicit* subtype
+of the protocol.
```

Comment:
Yes. I will add a note that type checkers should understand setting attributes to ``None`` in the sense of [data model docs](https://docs.python.org/3/reference/datamodel.html#special-method-names): "Setting a special method to ``None`` indicates that the corresponding operation is not available."